bundle/edu.gemini.itc.shared/src/main/java/edu/gemini/itc/shared/ConfigCreator.java

```diff
@@ -0,0 +1,201 @@
+    public static final ItemKey FPUKey = new ItemKey("instrument:fpu");
+    public static final ItemKey CcdXBinning = new ItemKey("instrument:ccdXBinning");
+    public static final ItemKey CcdYBinning = new ItemKey("instrument:ccdYBinning");
+    public static final ItemKey AmpGain = new ItemKey("instrument:gainChoice");
+    public static final ItemKey AmpCount = new ItemKey("instrument:ampCount");
+    public static final ItemKey DetectorManufacturerKey = new ItemKey("instrument:detectorManufacturer");
+    public static final ItemKey GuideWithPWFS2Key = new ItemKey("telescope:guideWithPWFS2");
+    public static final ItemKey TelescopeQKey = new ItemKey("telescope:q");
+    public static final ItemKey TelescopePKey = new ItemKey("telescope:p");
+    public static final ItemKey AOGuideStarTypeKey = new ItemKey("adaptive optics:guideStarType");
+    public static final ItemKey AOSystemKey = new ItemKey("adaptive optics:aoSystem");
+    public static final ItemKey AOFieldLensKey = new ItemKey("adaptive optics:fieldLens");
+
+    private final ItcParameters itcParams;
+    private final ObservationDetails obsDetailParams;
+
+    public ConfigCreator(final ItcParameters p) {
+        this.itcParams = p;
+        this.obsDetailParams = p.observation();
+    }
+
+    // create part of config common for all instruments
+    public Config[] createCommonConfig(int numberExposures) {
+        final CalculationMethod calcMethod = obsDetailParams.calculationMethod();
+        int numberCoadds = calcMethod.coaddsOrElse(1);
+        Config[] dc = new DefaultConfig[numberExposures];
+        String offset = Double.toString(calcMethod.offset());
+        // for spectroscopic observations we consider ABBA offset pattern
+        List<String> spectroscopyOffsets = new ArrayList<>(Arrays.asList("0", offset, offset, "0"));
+        // for imaging observations we consider ABAB offset pattern
+        List<String> imagingOffsets = new ArrayList<>(Arrays.asList("0", offset, "0", offset));
+        List<String> offsetList = new ArrayList<>();
+
+        for (int i = 0; i < (1 + numberExposures / 4); i++) {
+            if (calcMethod instanceof Imaging) {
+                offsetList.addAll(imagingOffsets);
+            } else if (calcMethod instanceof Spectroscopy) {
+                offsetList.addAll(spectroscopyOffsets);
+            }
+        }
+
+        for (int i = 0; i < dc.length; i++) {
+            Config step = new DefaultConfig();
+            dc[i] = step;
+
+            step.putItem(ExposureTimeKey, obsDetailParams.exposureTime());
+            step.putItem(ObserveTypeKey, "OBJECT");
+            step.putItem(ObserveClassKey, "SCIENCE");
+            step.putItem(CoaddsKey, numberCoadds);
+            step.putItem(TelescopePKey, "0");
+            step.putItem(TelescopeQKey, offsetList.get(i));
+
+            if (itcParams.telescope().getWFS().equals(GuideProbe.Type.PWFS)) {
+                step.putItem(GuideWithPWFS2Key, "guide");
+            }
```

Comment:
I'm not 100% sure about this. You continue to create the `DefaultConfig` with the same parameter values over and over again, with only different offsets.

What I would recommend is, before the `for` loop, creating the `DefaultConfig` with all the parameters that don't change.

Something like this, instead of line 77 - 91:
```java
final Config stepCommon = new DefaultConfig();
stepCommon.putItem(ExposureTimeKey, obsDetailsParams.exposureTime());
stepCommon.putItem(ObserveTypeKey, "OBJECT");
stepCommon.putItem(ObserveClassKey, "SCIENCE");
stepCommon.putItem(CoaddsKey, numberCoadds);
stepCommon.putItem(TelescopePKey, "0");

for (int i=0; i < dc.length; ++i) {
  final Config step = new DefaultConfig(stepCommon);
  step.putItem(TelescopeQKey, offset.get(i));

  if (itcParams.telescope().getWFS().equals(GuideProbe.Type.PWFS)) {
    step.putItem(GuideWithPWFS2Key, "guide");
  }

  dc[i] = step;
}
```

Reply:
Ok, got it. I'll change it (here and further) according to your suggestions, thank you for recommending an actual way of fixing the duplications. Something for me to keep in mind in future.